discovery/discovery_test.go

```diff
@@ -15,136 +15,568 @@ package discovery
-					isFirstSyncCall = false
-					initialGroups = tgs
+				currentCallGroup := make([]string, len(tgs))
+				for i, tg := range tgs {
+					currentCallGroup[i] = tg.Source
 				}
+				syncedGroups = append(syncedGroups, currentCallGroup)
 
-				if len(tgs) == len(expectedGroups) {
+				syncCallCount++
+				if syncCallCount == len(testCase.expectedSyncCalls) {
 					// All the groups are sent, we can start asserting.
-					wg.Done()
+					finalize <- true
 				}
 			},
 		})
 
 		ctx, cancel := context.WithCancel(context.Background())
 		defer cancel()
 
-		tp := newMockTargetProvider(updates)
+		tp := newMockTargetProvider(testCase.updates)
 		targetProviders := map[string]TargetProvider{}
 		targetProviders["testProvider"] = tp
 
 		go targetSet.Run(ctx)
 		targetSet.UpdateProviders(targetProviders)
 
-		finalize := make(chan struct{})
-		go func() {
-			defer close(finalize)
-			wg.Wait()
-		}()
-
 		select {
 		case <-time.After(20000 * time.Millisecond):
-			t.Errorf("In test case %v: Test timed out after 20000 millisecond. All targets should be sent within the timeout", i)
+			t.Errorf("In test case %v[%v]: Test timed out after 20000 millisecond. All targets should be sent within the timeout", i, testCase.title)
 
 		case <-finalize:
 
 			if *tp.callCount != 1 {
-				t.Errorf("In test case %v: TargetProvider Run should be called once only, was called %v times", i, *tp.callCount)
+				t.Errorf("In test case %v[%v]: TargetProvider Run should be called once only, was called %v times", i, testCase.title, *tp.callCount)
 			}
 
-			if len(updates) > 0 && updates[0].interval > 5000 {
-				// If the initial set of targets never arrive or arrive after 5 seconds.
-				// The first sync call should receive empty set of targets.
-				if len(initialGroups) != 0 {
-					t.Errorf("In test case %v: Expecting 0 initial target groups, received %v", i, len(initialGroups))
+			if len(syncedGroups) != len(testCase.expectedSyncCalls) {
+				t.Errorf("In test case %v[%v]: received sync calls: \n %v \n do not match expected calls: \n %v \n", i, testCase.title, syncedGroups, testCase.expectedSyncCalls)
+			}
+
+			for j := range syncedGroups {
+				if len(syncedGroups[j]) != len(testCase.expectedSyncCalls[j]) {
+					t.Errorf("In test case %v[%v]: received sync calls in call [%v]: \n %v \n do not match expected calls: \n %v \n", i, testCase.title, j, syncedGroups[j], testCase.expectedSyncCalls[j])
+				}
+
+				expectedGroupsMap := make(map[string]struct{})
+				for _, expectedGroup := range testCase.expectedSyncCalls[j] {
+					expectedGroupsMap[expectedGroup] = struct{}{}
 				}
+
+				for _, syncedGroup := range syncedGroups[j] {
+					if _, ok := expectedGroupsMap[syncedGroup]; ok == false {
+						t.Errorf("In test case %v[%v]: '%s' does not exist in expected target groups: %s", i, testCase.title, syncedGroup, testCase.expectedSyncCalls[j])
+					} else {
+						delete(expectedGroupsMap, syncedGroup) // Remove used targets from the map.
```

Comment:
Why? The map will be cleaned up by the GC anyway.

Reply:
This is not for cleaning up the the memory. It is for detecting the failures if the `sut` returns duplicate items. 
eg.
```
actualSyncedItems := []string{"tg1", "tg1"}
expectedSyncItems := map[string]struct{}{"tg1":... , "tg2":...}
```
Unless we delete `tg1` after it is used test would still pass.